flang/lib/Optimizer/HLFIR/IR/HLFIROps.cpp

```diff
@@ -207,20 +207,25 @@ static bool hasExplicitLowerBounds(mlir::Value shape) {
          mlir::isa<fir::ShapeShiftType, fir::ShiftType>(shape.getType());
 }
 
-static std::pair<mlir::Type, mlir::Value> updateDeclareInputTypeWithVolatility(
+static std::pair<mlir::Type, mlir::Value> updateDeclaredInputTypeWithVolatility(
     mlir::Type inputType, mlir::Value memref, mlir::OpBuilder &builder,
     fir::FortranVariableFlagsAttr fortran_attrs) {
   if (fortran_attrs &&
       bitEnumContainsAny(fortran_attrs.getFlags(),
                          fir::FortranVariableFlagsEnum::fortran_volatile)) {
+    // A volatile pointer's pointee is volatile.
     const bool isPointer = bitEnumContainsAny(
         fortran_attrs.getFlags(), fir::FortranVariableFlagsEnum::pointer);
+    // An allocatable's inner type's volatility matches that of the reference.
+    const bool isAllocatable = bitEnumContainsAny(
+        fortran_attrs.getFlags(), fir::FortranVariableFlagsEnum::allocatable);
```

Comment:
Line 213 checks if `fortran_attrs` is null before calling `bitEnumContainsAny`. Is that needed here too?

Reply:
We should have only entered this if-body if the attrs were non-null, but I think this is not very clear. Let me rewrite this a bit so it's more obvious. Thanks!